ResearchKit/Common/ORKTaskViewController.m

```diff
@@ -893,9 +893,18 @@ - (void)showViewController:(ORKStepViewController *)viewController goForward:(BO
         }
     }
     
-    if (step.identifier && ![_managedStepIdentifiers.lastObject isEqualToString:step.identifier]) {
-        [_managedStepIdentifiers addObject:step.identifier];
+    NSString *stepIdentifier = step.identifier;
+    if (stepIdentifier) {
+        NSUInteger indexOfStep = [_managedStepIdentifiers indexOfObject:stepIdentifier];
+        if (indexOfStep == NSNotFound) {
+            [_managedStepIdentifiers addObject:stepIdentifier];
+        } else {
+            // If the step being shown has been visited before, remove all the managed step identifiers visited after it
+            NSUInteger numberManagedOfStepIdentifiers = _managedStepIdentifiers.count;
+            [_managedStepIdentifiers removeObjectsInRange:NSMakeRange(indexOfStep + 1, numberManagedOfStepIdentifiers - (indexOfStep + 1))];
```

Comment:
I agree that loops are a bit controversial.

Originally, @jwe-apple and I agreed to allow loops and leave its usage as the developer responsibility. We log a warning to the console when a loop is detected: `Index of next step (\"%@\") is equal or lower than index of current step (\"%@\") in ordered task. Make sure this is intentional as you could loop idefinitely without appropriate navigation rules`.

So, do we still want to allow loops or not? I think it's better to have them for maximum flexibility. There might be some edge use cases for which they could be useful. Imagine a question like this: *"In light of this new piece of information, do you wish to change your answers?"*. If the user responds yes, the loop takes it back to a previous point in the survey.

If we do keep them, is it your contention that the `_managedStepIdentifiers` should hold all the history including loops? Aren't you worried about `managedResults` returning duplicate results for all the times the user looped over (even though all results point to the same result item)? Do we want to keep the whole backwards navigation stack including loops?